server/container_create.go

```diff
@@ -299,6 +300,28 @@ func generateUserString(username, imageUser string, uid *types.Int64Value) strin
 func (s *Server) CreateContainer(ctx context.Context, req *types.CreateContainerRequest) (res *types.CreateContainerResponse, retErr error) {
 	log.Infof(ctx, "Creating container: %s", translateLabelsToDescription(req.GetConfig().GetLabels()))
 
+	if req.Config.UserSpecifiedImage != "" {
+		systemCtx, err := s.contextForNamespace(req.SandboxConfig.Metadata.Namespace)
+		if err != nil {
+			return nil, fmt.Errorf("get context for namespace: %w", err)
+		}
+
+		img := req.Config.UserSpecifiedImage
+		log.Debugf(ctx, "Using pull policy path for image %s: %s", img, systemCtx.SignaturePolicyPath)
+
+		isAllowed, err := s.StorageImageServer().IsRunningImageAllowed(ctx, &systemCtx, img)
```

Comment:
This **must** pair the image ID used to work with the image with the `UserSpecifiedImage` value. Checking only one of these is not sufficient.

Reply:
If https://github.com/cri-o/cri-o/pull/7046#discussion_r1228433790 is correct, `IsRunningImageAllowed(…Image)` is not possible; and making two different signature checks on two possibly different objects seems not useful to me, or at least I can’t see the relationship.

This should:
- Resolve the `config.Image.Image` into a specific immutable c/storage image (either `storage.Image` object, or an `storage.Image.ID` value); or, maybe, to a specific manifest digest on a registry.
- Construct a `types.ImageSource` for exactly that immutable image.
- Validate the that immutable image, against the `UserSpecifiedImage`: use `UserSpecifiedImage` to look up the right policy scope, and verify the identity in the signature against `UserSpecifiedImage`
- Never do another `config.Image.Image` lookup (noting that `getRef` accepts names that can move); do all other operations involved in the container creation against the already-resolved immutable image identity.

Reply:
> > * Resolve the `config.Image.Image` into a specific immutable c/storage image (either `storage.Image` object, or an `storage.Image.ID` value); or, maybe, to a specific manifest digest on a registry.
> > * Construct a `types.ImageSource` for exactly that immutable image.
> 
> I've done this in https://github.com/cri-o/cri-o/pull/7046/files#diff-73f26dcabd7e433970fc0612ff0068d2d8f7cd61dda790bcd603195197aaad97R408-R418 via:

It’s not 100% clear to me that this will always resolve to the same image — nothing seems to be rejecting `configImage` values that use tags, at least within CRI-O. The lax parsing is not helping us here.

If the input were only an image ID, that would be clearly fine. If the input is a RepoDigests value, … maybe?

A tag would be a problem.

I’d rather prefer for this to be guaranteed to be immutable by the CRI-O codebase, not relying on Kubelet to never change what it submits (especially considering how underdocumented assumptions of the CRI API are).

---

>  > * Validate the that immutable image, against the `UserSpecifiedImage`: use `UserSpecifiedImage` to look up the right policy scope, and verify the identity in the signature against `UserSpecifiedImage`
> 
> I think I require more help on the API usage here. How can I move from the `UserSpecifiedImage` to a policy scope? Where is the link between the previously resolved store image and the resulting policy scope?

That’s one of the ways resolving an already-pulled image gets ”interesting” :)

Ordinarily, a signature verification during pull uses a `docker.Transport` implementation, so the signature is read from a registry and uses the policy section for `docker`.

This code would want to read an image from the c/storage transport, but still use a policy section for `docker`.

In practice, that means creating a new private `ImageReference` implementation, one that returns a `docker`-like values in `Transport()`, `PolicyConfigurationIdentity()` and `PolicyConfigurationNamespaces()`, but a `containers-storage` implementation of `NewImageSource`; and giving such an `ImageReference` implementation to `policyContext.IsRunningImageAllowed`. Probably use the existing `c/image/docker.{Parse,New}Reference`, and override the `NewImage*` methods.

> > * Never do another `config.Image.Image` lookup (noting that `getRef` accepts names that can move); do all other operations involved in the container creation against the already-resolved immutable image identity.
> 
> I think we only have lookups later for checkpoint-restore, which does an `ImageStatus` at some later point.

I would think the `config.Image.Image` must _somehow_ in actually creating the container. And that code path should also make sure to only refer to the already-resolved immutable identity. E.g. https://github.com/cri-o/cri-o/blob/4c8758b99741b5a0cf046271963ad759aeffe0be/server/container_create_linux.go#L229  is doing _some kind_ of lookup.

(I don’t immediately know whether that needs code changes deep in the container creation path, or imposing a strict format restriction on the input `config.Image.Image` field to be clearly immutable. We might want both.)

Reply:
Thank you for the insights @mtrmac! The resolving and especially the `ImageReference` implementation worries me with respect to complexity. I'm now thinking if it would be better to go one step back and do a plain image pull on container creation. The biggest question which comes into my mind would be: What would we do with the auth credentials? Would we have to pass them via the CRI? This would be a much bigger change compared to just passing the user specified image.

Reply:
The _implementation_ of `ImageReference` is really not that bad:
```go
type myReference {
   types.ImageReference
  storage types.ImageReference
}

func new(expectedIdentity string, local *store.Image) types.ImageReference {
    identity, … := docker.ParseReference("//"+expectedIdentity)
    storage, … := storage.NewStoreReference(…, nil, local.ID)
   return &myReference{ImageReference: identity, storage: storage}
}

func (r *myReference) NewImageSource() … {
   return r.storage.NewImageSource()
}
```

To me it seems tedious, not difficult.

---

I think the far more relevant tradeoff is that signatures on encrypted images don’t validate against already-pulled images. vs. that pulling on `CreateContainer` potentially changes semantics (if the reference is to a tag that is moved) _and_ significantly changes the performance profile of that operation.